File: server/src/main/java/org/dan/jadalnia/app/ws/WsBroadcast.java
package org.dan.jadalnia.app.ws;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.experimental.FieldDefaults;
import lombok.val;
import org.dan.jadalnia.app.festival.pojo.Fid;
import org.dan.jadalnia.app.order.pojo.Oid;
import org.dan.jadalnia.app.user.Uid;

import javax.inject.Inject;
import java.util.Collection;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
@RequiredArgsConstructor(onConstructor = @__(@Inject))
public class WsBroadcast {
    ObjectMapper objectMapper;
    Map<Fid, FestivalListeners> fid2Listeners;

    @SneakyThrows
    public void broadcast(Fid fid, MessageForClient message) {
        val listeners = getListeners(fid);
        val serializedMessage = objectMapper.writeValueAsBytes(message);

        broadcastTo(listeners.getCustomerListeners().values(), serializedMessage);
        broadcastTo(listeners.getUserListeners().values(), serializedMessage);
    }

    public FestivalListeners getListeners(Fid fid) {
        val listeners = fid2Listeners.get(fid);
        if (listeners == null) {
            fid2Listeners.putIfAbsent(
                    fid,
                    new FestivalListeners(
                            new ConcurrentHashMap<>(),
                            new ConcurrentHashMap<>()));
            return fid2Listeners.get(fid);
        }
        return listeners;
    }

    public Map<Oid, Uid> busyKelners(Fid fid) {
        return null;
    }

    public static void broadcastTo(
            Collection<? extends WsListener> listeners, byte[] message) {
        // close ws on failure
        listeners.forEach(listener -> listener.send(message));
    }
}

File: service/src/main/kotlin/org/dan/jadalnia/app/ws/WsBroadcast.kt
package org.dan.jadalnia.app.ws

import com.fasterxml.jackson.databind.ObjectMapper
import org.dan.jadalnia.app.festival.pojo.Festival

import org.dan.jadalnia.app.festival.pojo.Fid
import org.dan.jadalnia.app.user.Uid
import org.slf4j.LoggerFactory
import java.util.*

import javax.inject.Inject
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors.toList

class WsBroadcast @Inject constructor(
        val objectMapper: ObjectMapper,
        val fid2Listeners: MutableMap<Fid, FestivalListeners>) {

    fun broadcast(fid: Fid, message: MessageForClient) {
        val listeners = getListeners(fid)
        val serializedMessage: ByteArray = objectMapper.writeValueAsBytes(message)

        broadcastTo(listeners.customerListeners.values, serializedMessage)
        broadcastTo(listeners.volunteerListeners.values, serializedMessage)
    }

    fun notifyCustomers(fid: Fid, customers: Collection<Uid>, message: MessageForClient) {
        val listeners = getListeners(fid)
        val serializedMessage: ByteArray = objectMapper.writeValueAsBytes(message)
        log.info("notify customers {} of festival {} with {}", customers, fid, message)
        broadcastTo(
            customers.stream()
                .map { uid -> Optional.ofNullable(listeners.customerListeners[uid]) }
                .filter { o -> o.isPresent() }
                .map { o -> o.get() }
                .collect(toList()),
        serializedMessage)
    }

    fun broadcastTo(listeners: Collection<WsListener>, message: MessageForClient) {
        val serializedMessage: ByteArray = objectMapper.writeValueAsBytes(message)
        broadcastTo(listeners, serializedMessage)
    }

    fun getListeners(fid: Fid): FestivalListeners {
        val listeners = fid2Listeners[fid]
        if (listeners == null) {
            val festivalListeners = FestivalListeners(
                    customerListeners = ConcurrentHashMap(),
                    volunteerListeners = ConcurrentHashMap(),
                    kelnerUids = ConcurrentHashMap(),
                    kasierUids = ConcurrentHashMap(),
                    adminUids = ConcurrentHashMap(),
                    cookUids = ConcurrentHashMap())
            return fid2Listeners.putIfAbsent(fid, festivalListeners)
                    ?: return festivalListeners
        }
        return listeners
    }

    fun broadcastToFreeKelners(festival: Festival, message: MessageForClient) {
        broadcastTo(
                getListeners(festival.fid())
                        .volunteerListeners
                        .filterKeys { uid -> festival.freeKelners.containsKey(uid) }.values,
                message)
    }

    companion object {
        fun broadcastTo(
                listeners: Collection<WsListener>,
                message: ByteArray) {
            // close ws on failure
            listeners.forEach { listener -> listener.send(message) }
        }

        val log = LoggerFactory.getLogger(WsBroadcast::class.java)
    }
}
